gremlin-go/driver/connectionPool.go

```diff
@@ -0,0 +1,147 @@
+		err := connection.close()
+		if err != nil {
+			pool.logHandler.logf(Warning, errorClosingConnection, err.Error())
+		}
+	}
+}
+
+func (pool *loadBalancingPool) write(request *request) (ResultSet, error) {
+	connection, err := pool.getLeastUsedConnection()
+	if err != nil {
+		return nil, err
+	}
+	return connection.write(request)
+}
+
+func (pool *loadBalancingPool) getLeastUsedConnection() (*connection, error) {
+	pool.loadBalanceLock.Lock()
+	defer pool.loadBalanceLock.Unlock()
+	if len(pool.connections) == 0 {
+		return pool.newConnection()
+	} else {
+		var leastUsed *connection = nil
+		validIndex := 0
+		for _, connection := range pool.connections {
+			// Purge dead connections from pool
+			if connection.state == established {
+				// Close and purge connections from pool if there is more than one being unused
+				if leastUsed != nil && (leastUsed.activeResults() == 0 && connection.activeResults() == 0) {
+					// Close the connection asynchronously since it is a high-latency method
+					go func() {
+						pool.logHandler.log(Debug, closeUnusedPoolConnection)
+						err := connection.close()
+						if err != nil {
+							pool.logHandler.logf(Warning, errorClosingConnection, err.Error())
+						}
+					}()
+
+					continue
+				}
+
+				// Mark connection as valid to keep
+				pool.connections[validIndex] = connection
+				validIndex++
+
+				// Set the least used connection
+				if leastUsed == nil || connection.activeResults() < leastUsed.activeResults() {
+					leastUsed = connection
+				}
+			} else {
+				pool.logHandler.log(Debug, purgingDeadConnection)
+			}
+		}
+
+		// Deallocate truncated dead connections to prevent memory leak
+		for invalidIndex := validIndex; invalidIndex < len(pool.connections); invalidIndex++ {
+			pool.connections[invalidIndex] = nil
+		}
+		pool.connections = pool.connections[:validIndex]
+
+		// Create new connection if no valid connections were found in the pool or the least used connection exceeded
+		// the concurrent usage threshold while the pool still has capacity for a new connection
+		if leastUsed == nil ||
+			(leastUsed.activeResults() >= pool.newConnectionThreshold && len(pool.connections) < cap(pool.connections)) {
+			return pool.newConnection()
+		} else {
+			return leastUsed, nil
+		}
+	}
+}
+
+func (pool *loadBalancingPool) newConnection() (*connection, error) {
+	connection, err := createConnection(pool.url, pool.logHandler, pool.connSettings)
+	if err != nil {
+		return nil, err
+	}
+	pool.connections = append(pool.connections, connection)
+	return connection, nil
+}
+
+func newLoadBalancingPool(url string, logHandler *logHandler, connSettings *connectionSettings, newConnectionThreshold int,
+	maximumConcurrentConnections int) (connectionPool, error) {
+	pool := make([]*connection, 0, maximumConcurrentConnections)
+	initialConnection, err := createConnection(url, logHandler, connSettings)
+	if err != nil {
+		return nil, err
+	}
+	pool = append(pool, initialConnection)
```

Comment:
Code duplication. Refactor `newConnection()` to work for this use case as well instead?